File: .UNSAFE-ALL-BINDINGS-DELETED-FINAL-20260125/bindings/clojure/src/proven/safe_uuid.clj
;; SPDX-License-Identifier: PMPL-1.0
;; SPDX-FileCopyrightText: 2025 Hyperpolymath

(ns proven.safe-uuid
  "Safe UUID generation and validation following RFC 4122."
  (:require [clojure.string :as str])
  (:import [java.util UUID]))

;; UUID version types
(def uuid-versions
  "UUID version identifiers."
  {:v1 1  ; Time-based
   :v2 2  ; DCE Security
   :v3 3  ; Name-based (MD5)
   :v4 4  ; Random
   :v5 5  ; Name-based (SHA-1)
   :nil 0})

;; UUID variant types
(def uuid-variants
  "UUID variant identifiers."
  {:ncs 0
   :rfc4122 1
   :microsoft 2
   :future 3})

;; Namespace UUIDs per RFC 4122
(def namespace-dns
  "DNS namespace UUID."
  {:bytes [0x6b 0xa7 0xb8 0x10 0x9d 0xad 0x11 0xd1
           0x80 0xb4 0x00 0xc0 0x4f 0xd4 0x30 0xc8]})

(def namespace-url
  "URL namespace UUID."
  {:bytes [0x6b 0xa7 0xb8 0x11 0x9d 0xad 0x11 0xd1
           0x80 0xb4 0x00 0xc0 0x4f 0xd4 0x30 0xc8]})

(def namespace-oid
  "OID namespace UUID."
  {:bytes [0x6b 0xa7 0xb8 0x12 0x9d 0xad 0x11 0xd1
           0x80 0xb4 0x00 0xc0 0x4f 0xd4 0x30 0xc8]})

(def namespace-x500
  "X.500 namespace UUID."
  {:bytes [0x6b 0xa7 0xb8 0x14 0x9d 0xad 0x11 0xd1
           0x80 0xb4 0x00 0xc0 0x4f 0xd4 0x30 0xc8]})

(def nil-uuid
  "The nil UUID (all zeros)."
  {:bytes (vec (repeat 16 0))})

(defrecord SafeUuid [bytes])

(defn- byte-to-hex
  "Convert a byte to two hex characters."
  [byte-value]
  (format "%02x" (bit-and byte-value 0xFF)))

(defn- hex-char-to-nibble
  "Convert a hex character to its nibble value."
  [hex-char]
  (let [character-code (int hex-char)]
    (cond
      (<= (int \0) character-code (int \9)) (- character-code (int \0))
      (<= (int \a) character-code (int \f)) (+ 10 (- character-code (int \a)))
      (<= (int \A) character-code (int \F)) (+ 10 (- character-code (int \A)))
      :else nil)))

(defn from-bytes
  "Create a SafeUuid from a 16-byte vector."
  [byte-vector]
  (when (and (vector? byte-vector) (= 16 (count byte-vector)))
    (->SafeUuid byte-vector)))

(defn get-bytes
  "Get the raw bytes of a UUID."
  [uuid]
  (:bytes uuid))

(defn get-version
  "Get the UUID version."
  [uuid]
  (let [version-byte (nth (:bytes uuid) 6)
        version-nibble (bit-and (bit-shift-right version-byte 4) 0x0F)]
    (case version-nibble
      1 :v1
      2 :v2
      3 :v3
      4 :v4
      5 :v5
      :nil)))

(defn get-variant
  "Get the UUID variant."
  [uuid]
  (let [variant-byte (nth (:bytes uuid) 8)]
    (cond
      (zero? (bit-and variant-byte 0x80)) :ncs
      (= 0x80 (bit-and variant-byte 0xC0)) :rfc4122
      (= 0xC0 (bit-and variant-byte 0xE0)) :microsoft
      :else :future)))

(defn nil?-uuid
  "Check if the UUID is the nil UUID."
  [uuid]
  (every? zero? (:bytes uuid)))

(defn format-canonical
  "Format UUID as canonical string (8-4-4-4-12)."
  [uuid]
  (let [bytes-vec (:bytes uuid)]
    (str
     (apply str (map byte-to-hex (subvec bytes-vec 0 4)))
     "-"
     (apply str (map byte-to-hex (subvec bytes-vec 4 6)))
     "-"
     (apply str (map byte-to-hex (subvec bytes-vec 6 8)))
     "-"
     (apply str (map byte-to-hex (subvec bytes-vec 8 10)))
     "-"
     (apply str (map byte-to-hex (subvec bytes-vec 10 16))))))

(defn format-urn
  "Format UUID as URN (urn:uuid:...)."
  [uuid]
  (str "urn:uuid:" (format-canonical uuid)))

(defn format-braced
  "Format UUID with braces ({...})."
  [uuid]
  (str "{" (format-canonical uuid) "}"))

(defn format-compact
  "Format UUID without hyphens."
  [uuid]
  (apply str (map byte-to-hex (:bytes uuid))))

(defn parse
  "Parse a UUID from canonical string format.
   Returns {:ok SafeUuid} or {:error message}."
  [uuid-string]
  (cond
    (str/blank? uuid-string)
    {:error "empty_uuid_string"}

    (not= 36 (count uuid-string))
    {:error "invalid_uuid_length"}

    (or (not= \- (nth uuid-string 8))
        (not= \- (nth uuid-string 13))
        (not= \- (nth uuid-string 18))
        (not= \- (nth uuid-string 23)))
    {:error "invalid_uuid_format"}

    :else
    (let [hex-string (str/replace uuid-string "-" "")]
      (if (not= 32 (count hex-string))
        {:error "invalid_uuid_hex_length"}
        (let [hex-chars (seq hex-string)
              nibbles (map hex-char-to-nibble hex-chars)]
          (if (some nil? nibbles)
            {:error "invalid_hex_character"}
            (let [byte-values (mapv (fn [[high-nibble low-nibble]]
                                      (bit-or (bit-shift-left high-nibble 4) low-nibble))
                                    (partition 2 nibbles))]
              {:ok (->SafeUuid byte-values)})))))))

(defn parse!
  "Parse a UUID from string. Throws ex-info on error."
  [uuid-string]
  (let [result (parse uuid-string)]
    (if (:ok result)
      (:ok result)
      (throw (ex-info (str "Invalid UUID: " (:error result))
                      {:type :invalid-uuid
                       :error (:error result)
                       :input uuid-string})))))

(defn valid?
  "Check if a string is a valid UUID format."
  [uuid-string]
  (contains? (parse uuid-string) :ok))

(defn generate-v4
  "Generate a random v4 UUID using Java's UUID generator."
  []
  (let [java-uuid (UUID/randomUUID)
        most-significant-bits (.getMostSignificantBits java-uuid)
        least-significant-bits (.getLeastSignificantBits java-uuid)
        bytes-vec (vec (concat
                        (for [shift-amount [56 48 40 32 24 16 8 0]]
                          (bit-and (bit-shift-right most-significant-bits shift-amount) 0xFF))
                        (for [shift-amount [56 48 40 32 24 16 8 0]]
                          (bit-and (bit-shift-right least-significant-bits shift-amount) 0xFF))))]
    (->SafeUuid bytes-vec)))

(defn v4-from-bytes
  "Create a v4 UUID from 16 random bytes (sets version and variant bits)."
  [random-bytes]
  (when (and (sequential? random-bytes) (= 16 (count random-bytes)))
    (let [mutable-bytes (vec random-bytes)
          ;; Set version to 4
          byte-6 (bit-or (bit-and (nth mutable-bytes 6) 0x0F) 0x40)
          ;; Set variant to RFC 4122
          byte-8 (bit-or (bit-and (nth mutable-bytes 8) 0x3F) 0x80)]
      (->SafeUuid (-> mutable-bytes
                      (assoc 6 byte-6)
                      (assoc 8 byte-8))))))

(defn equal?
  "Check if two UUIDs are equal."
  [uuid-a uuid-b]
  (= (:bytes uuid-a) (:bytes uuid-b)))

(defn compare-uuids
  "Compare two UUIDs lexicographically."
  [uuid-a uuid-b]
  (compare (:bytes uuid-a) (:bytes uuid-b)))

(defn to-java-uuid
  "Convert SafeUuid to java.util.UUID."
  [uuid]
  (UUID/fromString (format-canonical uuid)))

(defn from-java-uuid
  "Convert java.util.UUID to SafeUuid."
  [java-uuid]
  (parse (.toString java-uuid)))

File: .UNSAFE-ALL-BINDINGS-DELETED-FINAL-20260125/bindings/php/src/Proven/SafeUuid.php
<?php
// SPDX-License-Identifier: AGPL-3.0-or-later
// SPDX-FileCopyrightText: 2025 Hyperpolymath

declare(strict_types=1);

namespace Proven;

/**
 * UUID version types as defined by RFC 4122 and RFC 9562.
 */
enum UuidVersion: int
{
    case V1 = 1;  // Time-based
    case V3 = 3;  // Name-based (MD5)
    case V4 = 4;  // Random
    case V5 = 5;  // Name-based (SHA-1)
    case V6 = 6;  // Reordered time-based
    case V7 = 7;  // Unix epoch time-based
    case V8 = 8;  // Custom

    /**
     * Get a human-readable description of this version.
     *
     * @return string Description of the UUID version
     */
    public function description(): string
    {
        return match ($this) {
            self::V1 => 'Time-based (MAC address)',
            self::V3 => 'Name-based (MD5 hash)',
            self::V4 => 'Random',
            self::V5 => 'Name-based (SHA-1 hash)',
            self::V6 => 'Reordered time-based',
            self::V7 => 'Unix epoch time-based',
            self::V8 => 'Custom',
        };
    }
}

/**
 * UUID variant types as defined by RFC 4122.
 */
enum UuidVariant: int
{
    case NCS = 0;           // Reserved for NCS backward compatibility
    case RFC4122 = 1;       // RFC 4122 (standard)
    case Microsoft = 2;     // Reserved for Microsoft backward compatibility
    case Future = 3;        // Reserved for future definition

    /**
     * Get a human-readable description of this variant.
     *
     * @return string Description of the UUID variant
     */
    public function description(): string
    {
        return match ($this) {
            self::NCS => 'Reserved (NCS backward compatibility)',
            self::RFC4122 => 'RFC 4122 (standard)',
            self::Microsoft => 'Reserved (Microsoft backward compatibility)',
            self::Future => 'Reserved (future definition)',
        };
    }
}

/**
 * Immutable UUID value object with safe parsing and formatting.
 */
readonly class Uuid
{
    /** @var string The 16-byte binary representation */
    private string $bytes;

    /**
     * Create a UUID from 16 raw bytes.
     *
     * @param string $bytes 16-byte binary string
     * @throws \InvalidArgumentException If bytes is not exactly 16 bytes
     */
    private function __construct(string $bytes)
    {
        if (strlen($bytes) !== 16) {
            throw new \InvalidArgumentException('UUID must be exactly 16 bytes');
        }
        $this->bytes = $bytes;
    }

    /**
     * Parse a UUID from a string representation.
     *
     * Accepts formats:
     * - Standard: xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx
     * - Compact: xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
     * - URN: urn:uuid:xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx
     * - Braced: {xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx}
     *
     * @param string $uuidString The UUID string to parse
     * @return self|null The parsed UUID or null if invalid
     */
    public static function parse(string $uuidString): ?self
    {
        $input = trim($uuidString);

        // Handle URN format
        if (str_starts_with(strtolower($input), 'urn:uuid:')) {
            $input = substr($input, 9);
        }

        // Handle braced format
        if (str_starts_with($input, '{') && str_ends_with($input, '}')) {
            $input = substr($input, 1, -1);
        }

        // Remove hyphens for compact parsing
        $hex = str_replace('-', '', $input);

        // Validate hex string
        if (strlen($hex) !== 32 || !ctype_xdigit($hex)) {
            return null;
        }

        $bytes = hex2bin($hex);
        if ($bytes === false || strlen($bytes) !== 16) {
            return null;
        }

        return new self($bytes);
    }

    /**
     * Create a UUID from 16 raw bytes.
     *
     * @param string $bytes 16-byte binary string
     * @return self|null The UUID or null if invalid length
     */
    public static function fromBytes(string $bytes): ?self
    {
        if (strlen($bytes) !== 16) {
            return null;
        }
        return new self($bytes);
    }

    /**
     * Generate a random version 4 UUID.
     *
     * @return self A new random UUID
     */
    public static function v4(): self
    {
        $bytes = random_bytes(16);

        // Set version to 4 (random)
        $bytes[6] = chr((ord($bytes[6]) & 0x0F) | 0x40);

        // Set variant to RFC 4122
        $bytes[8] = chr((ord($bytes[8]) & 0x3F) | 0x80);

        return new self($bytes);
    }

    /**
     * Generate a version 7 UUID (Unix epoch time-based).
     *
     * @return self A new time-based UUID
     */
    public static function v7(): self
    {
        // Get current time in milliseconds
        $timestampMs = (int)(microtime(true) * 1000);

        // Pack timestamp into first 6 bytes (48 bits)
        $bytes = pack('J', $timestampMs);
        $bytes = substr($bytes, 2, 6); // Take last 6 bytes (48 bits)

        // Add 10 random bytes
        $bytes .= random_bytes(10);

        // Set version to 7
        $bytes[6] = chr((ord($bytes[6]) & 0x0F) | 0x70);

        // Set variant to RFC 4122
        $bytes[8] = chr((ord($bytes[8]) & 0x3F) | 0x80);

        return new self($bytes);
    }

    /**
     * Create the nil UUID (all zeros).
     *
     * @return self The nil UUID
     */
    public static function nil(): self
    {
        return new self(str_repeat("\x00", 16));
    }

    /**
     * Create the max UUID (all ones).
     *
     * @return self The max UUID
     */
    public static function max(): self
    {
        return new self(str_repeat("\xFF", 16));
    }

    /**
     * Format the UUID as a standard string.
     *
     * @return string The UUID in format xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx
     */
    public function format(): string
    {
        $hex = bin2hex($this->bytes);
        return sprintf(
            '%s-%s-%s-%s-%s',
            substr($hex, 0, 8),
            substr($hex, 8, 4),
            substr($hex, 12, 4),
            substr($hex, 16, 4),
            substr($hex, 20, 12)
        );
    }

    /**
     * Format the UUID as a compact string (no hyphens).
     *
     * @return string The UUID as 32 hex characters
     */
    public function formatCompact(): string
    {
        return bin2hex($this->bytes);
    }

    /**
     * Format the UUID as a URN.
     *
     * @return string The UUID as urn:uuid:xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx
     */
    public function formatUrn(): string
    {
        return 'urn:uuid:' . $this->format();
    }

    /**
     * Format the UUID in Microsoft braced format.
     *
     * @return string The UUID as {xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx}
     */
    public function formatBraced(): string
    {
        return '{' . $this->format() . '}';
    }

    /**
     * Get the raw 16-byte binary representation.
     *
     * @return string 16-byte binary string
     */
    public function toBytes(): string
    {
        return $this->bytes;
    }

    /**
     * Get the UUID version.
     *
     * @return UuidVersion|null The version or null if unknown
     */
    public function getVersion(): ?UuidVersion
    {
        $versionNibble = (ord($this->bytes[6]) >> 4) & 0x0F;
        return UuidVersion::tryFrom($versionNibble);
    }

    /**
     * Get the UUID variant.
     *
     * @return UuidVariant The variant
     */
    public function getVariant(): UuidVariant
    {
        $variantByte = ord($this->bytes[8]);

        if (($variantByte & 0x80) === 0x00) {
            return UuidVariant::NCS;
        }
        if (($variantByte & 0xC0) === 0x80) {
            return UuidVariant::RFC4122;
        }
        if (($variantByte & 0xE0) === 0xC0) {
            return UuidVariant::Microsoft;
        }
        return UuidVariant::Future;
    }

    /**
     * Check if this is the nil UUID.
     *
     * @return bool True if all bytes are zero
     */
    public function isNil(): bool
    {
        return $this->bytes === str_repeat("\x00", 16);
    }

    /**
     * Check if this is the max UUID.
     *
     * @return bool True if all bytes are 0xFF
     */
    public function isMax(): bool
    {
        return $this->bytes === str_repeat("\xFF", 16);
    }

    /**
     * Compare two UUIDs for equality.
     *
     * @param self $other The UUID to compare with
     * @return bool True if equal
     */
    public function equals(self $other): bool
    {
        return hash_equals($this->bytes, $other->bytes);
    }

    /**
     * Compare two UUIDs for ordering.
     *
     * @param self $other The UUID to compare with
     * @return int -1, 0, or 1
     */
    public function compare(self $other): int
    {
        return strcmp($this->bytes, $other->bytes) <=> 0;
    }

    /**
     * Get the string representation.
     *
     * @return string The formatted UUID
     */
    public function __toString(): string
    {
        return $this->format();
    }
}

/**
 * Safe UUID operations with validation and parsing utilities.
 */
class SafeUuid
{
    /**
     * Check if a string is a valid UUID.
     *
     * @param string $uuidString The string to validate
     * @return bool True if valid UUID
     */
    public static function isValid(string $uuidString): bool
    {
        return Uuid::parse($uuidString) !== null;
    }

    /**
     * Parse a UUID string, returning null on failure.
     *
     * @param string $uuidString The string to parse
     * @return Uuid|null The parsed UUID or null
     */
    public static function parse(string $uuidString): ?Uuid
    {
        return Uuid::parse($uuidString);
    }

    /**
     * Generate a new random UUID (v4).
     *
     * @return Uuid A new random UUID
     */
    public static function generate(): Uuid
    {
        return Uuid::v4();
    }

    /**
     * Generate a new time-based UUID (v7).
     *
     * @return Uuid A new time-based UUID
     */
    public static function generateV7(): Uuid
    {
        return Uuid::v7();
    }

    /**
     * Normalize a UUID string to standard format.
     *
     * @param string $uuidString The UUID string to normalize
     * @return string|null The normalized UUID or null if invalid
     */
    public static function normalize(string $uuidString): ?string
    {
        $uuid = Uuid::parse($uuidString);
        return $uuid?->format();
    }

    /**
     * Compare two UUID strings for equality (case-insensitive).
     *
     * @param string $uuidStringA First UUID string
     * @param string $uuidStringB Second UUID string
     * @return bool True if equal
     */
    public static function equals(string $uuidStringA, string $uuidStringB): bool
    {
        $uuidA = Uuid::parse($uuidStringA);
        $uuidB = Uuid::parse($uuidStringB);

        if ($uuidA === null || $uuidB === null) {
            return false;
        }

        return $uuidA->equals($uuidB);
    }
}
